site/app/libraries/database/DatabaseQueries.php

```diff
@@ -2909,4 +2910,226 @@ public function updateGradeableNew(\app\models\gradeable\Gradeable $gradeable) {
+     * Updates an existing graded component in the database
+     * @param TaGradedGradeable $ta_graded_gradeable The ta graded gradeable that $graded_component belongs to
+     * @param GradedComponent $graded_component
+     */
+    private function updateGradedComponent(TaGradedGradeable $ta_graded_gradeable, GradedComponent $graded_component) {
+        if ($graded_component->isModified()) {
+            $params = [
+                $graded_component->getScore(),
+                $graded_component->getComment(),
+                $graded_component->getGradedVersion(),
+                DateUtils::dateTimeToString($graded_component->getGradeTime()),
+                $ta_graded_gradeable->getId(),
+                $graded_component->getComponentId(),
+                $graded_component->getGraderId()
+            ];
+            $query = "
+                UPDATE gradeable_component_data SET 
+                  gcd_score=?,
+                  gcd_component_comment=?,
+                  gcd_graded_version=?,
+                  gcd_grade_time=?
+                WHERE gd_id=? AND gc_id=? AND gcd_grader_id=?";
+            $this->course_db->query($query, $params);
+        }
+    }
+
+    /**
+     * Update/create the components/marks for a gradeable.
+     *  Note: This does NOT delete components.  Any missing components will left alone
+     * @param TaGradedGradeable $ta_graded_gradeable
+     */
+    private function updateGradedComponents(TaGradedGradeable $ta_graded_gradeable) {
+        // iterate through graded components and see if any need updating/creating
+        /** @var GradedComponent[] $component_grades */
+        foreach ($ta_graded_gradeable->getGradedComponents() as $component_grades) {
+            foreach ($component_grades as $component_grade) {
+                // This means the component wasn't loaded from the database, ergo its new
+                if ($component_grade->getDbMarkIds() === null) {
+                    $this->createGradedComponent($ta_graded_gradeable, $component_grade);
+                } else {
+                    $this->updateGradedComponent($ta_graded_gradeable, $component_grade);
+                }
+
+                // If the marks have been modified, this means we need to update the entries
+                if ($component_grade->isMarksModified()) {
+                    $new_marks = array_diff($component_grade->getMarkIds(), $component_grade->getDbMarkIds() ?? []);
+                    $deleted_marks = array_diff($component_grade->getDbMarkIds() ?? [], $component_grade->getMarkIds());
+                    $this->deleteGradedComponentMarks(
+                        $ta_graded_gradeable,
+                        $component_grade,
+                        $deleted_marks);
+                    $this->createGradedComponentMarks(
+                        $ta_graded_gradeable,
+                        $component_grade,
+                        $new_marks);
+                }
+            }
+        }
```

Comment:
UpdateGradedComponents doesn't delete removed components from the database, and there doesn't seem to be a method for that.

Reply:
I couldn't think of a use case where you'd want to delete the components without deleting the whole TA grade, but i'll add the feature for completeness, since I'm probably not thinking about everything.

Reply:
If you ungrade only one component it removes the GradedComponent (at least visually) but there can still be other GradedComponents

Reply:
This got marked as outdated, but its still relevant.  I'll add this in.